drivers/bme680/bme680_common.c

```diff
@@ -0,0 +1,166 @@
+    var2 = (dev->calib.par_g1 + 784) * (((((dev->calib.par_g2 + 154009) * (dev->params.gas_heating_temp) * 5) / 100) + 3276800) / 10);
+    var3 = var1 + (var2 / 2);
+    var4 = (var3 / ((dev->calib.res_heat_range) + 4));
+    var5 = (131 * dev->calib.res_heat_val) + 65536;
+    heatr_res_x100 = (int32_t) (((var4 / var5) - 250) * 34);
+    heatr_res = (uint8_t) ((heatr_res_x100 + 50) / 100);
+
+    *res = heatr_res;
+
+    return 0;
+}
+
+/* calculate compensated temperature using data sheet formula */
+int calc_temp(const bme680_common_t *dev, uint32_t *res, uint32_t *t_fine, uint32_t temp_adc)
+{
+    int32_t var1 = ((int32_t)temp_adc >> 3) - ((int32_t)dev->calib.par_t1 << 1);
+    int32_t var2 = (var1 * (int32_t)dev->calib.par_t2) >> 11;
+    int32_t var3 = ((((var1 >> 1) * (var1 >> 1)) >> 12) * ((int32_t)dev->calib.par_t3 << 4)) >> 14;
+    *t_fine = var2 + var3;
+    int32_t temp_comp = ((*t_fine * 5) + 128) >> 8;
+
+    if ((temp_comp < (-4000)) | (temp_comp > 8500)){
+        return -ERANGE;
+    }
+
+    *res = temp_comp;
+
+    return 0;
+}
+
+/* calculate humdity using datasheet formula */
+int calc_hum(const bme680_common_t *dev, uint32_t *res, const uint32_t *temp_comp, uint16_t hum_adc)
+{
+    int32_t var1 = (int32_t) (hum_adc - ((int32_t) ((int32_t) dev->calib.par_h1 << 4)))
+        - (((((int32_t) *temp_comp) * (int32_t) dev->calib.par_h3) / ((int32_t) 100)) >> 1);
+    int32_t var2 = ((int32_t) dev->calib.par_h2
+        * ((((*temp_comp) * (int32_t) dev->calib.par_h4) / ((int32_t) 100))
+            + (((((int32_t) *temp_comp) * ((((int32_t) *temp_comp) * (int32_t) dev->calib.par_h5) / ((int32_t) 100))) >> 6)
+                / ((int32_t) 100)) + (int32_t) (1 << 14))) >> 10;
+    int32_t var3 = var1 * var2;
+    int32_t var4 = (int32_t) dev->calib.par_h6 << 7;
+    var4 = ((var4) + ((((int32_t) *temp_comp) * (int32_t) dev->calib.par_h7) / ((int32_t) 100))) >> 4;
+    int32_t var5 = ((var3 >> 14) * (var3 >> 14)) >> 10;
+    int32_t var6 = (var4 * var5) >> 1;
+    int32_t hum_comp = (((var3 + var6) >> 10) * ((int32_t) 1000)) >> 12;
+
+    if ((hum_comp < 0) || (hum_comp > 100000)){
+        return -ERANGE;
+    }
+
+    *res = (uint32_t) hum_comp;
+    return 0;
+}
+
+/* translate heating duration to register value */
+void calc_heater_dur(uint16_t dur, uint8_t *res)
+{
+    uint8_t factor = 0;
+
+    if (dur >= 0xfc0) {
+        *res = 0xff; /* Max duration */
+    } else {
+        while (dur > 0x3F) {
+            dur = dur / 4;
+            factor += 1;
+        }
+        *res = (uint8_t) (dur + (factor * 64));
+    }
+}
+
+/* calculate pressure using datasheet formula */
+int calc_press(const bme680_common_t *dev, uint32_t *res, uint32_t *t_fine, uint32_t press_adc)
+{
+    int32_t var1, var2, var3, pressure_comp;
+
+    var1 = (((int32_t)* t_fine) >> 1) - 64000;
+    var2 = ((((var1 >> 2) * (var1 >> 2)) >> 11) * (int32_t)dev->calib.par_p6) >> 2;
```

Comment:
```suggestion
    var2 = ((((var1 >> 2) * (var1 >> 2)) >> 11) * dev->calib.par_p6) >> 2;
```

If you apply a binary operator on two variables of different integer type, the type with the highest range is chosen and the other type is converted to it prior to the calculation. E.g. in

```
int32_t a;
int8_t b;
int32_t c = a * b; // <- Here b gets implicitly converted to `int32_t` to match the type of a before the multiplication is carried out
```